### ISSUE TITLE
Error: Call to a member function onMessage() on null

### ISSUE THREAD

**Issue by ikovalyov** [ratchetphp/Ratchet#557]
Hi.
We found strange error in our server log:

```
Error: Call to a member function onMessage() on null
1 File "/home/api.perfops.net/vendor/cboden/ratchet/src/Ratchet/Http/HttpServer.php" line 54 in <main>
2 File "/home/api.perfops.net/vendor/cboden/ratchet/src/Ratchet/Server/IoServer.php" line 112 in onMessage
3 File "/home/api.perfops.net/vendor/cboden/ratchet/src/Ratchet/Server/IoServer.php" line 95 in onMessage
4 File "/home/api.perfops.net/vendor/evenement/evenement/src/Evenement/EventEmitterTrait.php" line 70 in handleData
5 File "/home/api.perfops.net/vendor/react/stream/src/Util.php" line 41 in Ratchet\Server\{closure}
6 File "/home/api.perfops.net/vendor/evenement/evenement/src/Evenement/EventEmitterTrait.php" line 70 in emit
7 File "/home/api.perfops.net/vendor/react/stream/src/Stream.php" line 173 in React\Stream\{closure}
8 File "<internal>" line (unknown) in emit
9 File "/home/api.perfops.net/vendor/react/event-loop/src/StreamSelectLoop.php" line 236 in handleData
10 File "/home/api.perfops.net/vendor/react/event-loop/src/StreamSelectLoop.php" line 205 in call_user_func
11 File "/home/api.perfops.net/vendor/cboden/ratchet/src/Ratchet/Server/IoServer.php" line 74 in waitForStreamActivity
12 File "/home/api.perfops.net/vendor/cboden/ratchet/src/Ratchet/App.php" line 145 in run
...
```
It looks like _httpServer was not set for some reason for connection.

Also little different but similar issue:

```
E_NOTICE: Undefined property: Ratchet\Server\IoConnection::$controller: Undefined property: Ratchet\Server\IoConnection::$controller
1 File "/home/api.perfops.net/vendor/prospect-one/yassa-rollbar/src/Yassa/Rollbar/RollbarNotifier.php" line 66 in <main>
2 File "/home/api.perfops.net/vendor/cboden/ratchet/src/Ratchet/Http/Router.php" line 71 in errorHandler
3 File "/home/api.perfops.net/vendor/cboden/ratchet/src/Ratchet/Http/HttpServer.php" line 54 in report_php_error
4 File "/home/api.perfops.net/vendor/cboden/ratchet/src/Ratchet/Server/IoServer.php" line 112 in onMessage
5 File "/home/api.perfops.net/vendor/cboden/ratchet/src/Ratchet/Server/IoServer.php" line 95 in onMessage
6 File "/home/api.perfops.net/vendor/evenement/evenement/src/Evenement/EventEmitterTrait.php" line 70 in handleData
7 File "/home/api.perfops.net/vendor/react/stream/src/Util.php" line 41 in Ratchet\Server\{closure}
8 File "/home/api.perfops.net/vendor/evenement/evenement/src/Evenement/EventEmitterTrait.php" line 70 in emit
9 File "/home/api.perfops.net/vendor/react/stream/src/Stream.php" line 173 in React\Stream\{closure}
10 File "<internal>" line (unknown) in emit
11 File "/home/api.perfops.net/vendor/react/event-loop/src/StreamSelectLoop.php" line 236 in handleData
12 File "/home/api.perfops.net/vendor/react/event-loop/src/StreamSelectLoop.php" line 205 in call_user_func
13 File "/home/api.perfops.net/vendor/cboden/ratchet/src/Ratchet/Server/IoServer.php" line 74 in waitForStreamActivity
14 File "/home/api.perfops.net/vendor/cboden/ratchet/src/Ratchet/App.php" line 145 in run
...
```

It looks like controller is empty. It catches this error then trying to send error message but _httpServer is empty as well. So it dies.
Do you have an idea what caused it and how we can make sure it won't happen again?
Thanks.

**Comment by cboden** [comment]
Can you post your code please?

**Comment by ikovalyov** [comment]
Full project logic is very complex and I won't be able to isolate something which is causing this issue as this project is running for few months and this error appeared only 1 time. Here are some code samples:

initializing app:
```
        $app = new App($host, $port, '0.0.0.0', $loop);
        foreach($routes as $route) {
            if (isset($route['factory'])) {
                $obj = $container->get($route['factory']);
            } else {
                $obj = new $route['class'];
            }
            $app->route($route['route'], $obj, $route['params']);
        }
        return $app;
```

All controllers are successors of this class:

```
abstract class AbstractMessageComponent implements MessageComponentInterface
{
    /**
     * @var ConnectionInterface[]|SplObjectStorage
     */
    protected $clients;

    /**
     * AbstractMessageComponent constructor.
     */
    public function __construct() {
        $this->clients = new SplObjectStorage;
    }

    /**
     * @param ConnectionInterface $conn
     * @param RequestInterface|null $request
     */
    function onOpen(ConnectionInterface $conn, RequestInterface $request = null)
    {
        $this->clients->attach($conn);
    }

    /**
     * This is called before or after a socket is closed (depends on how it's closed).  SendMessage to $conn will not result in an error if it has already been closed.
     * @param  ConnectionInterface $conn The socket/connection that is closing/closed
     * @throws \Exception
     */
    function onClose(ConnectionInterface $conn)
    {
        $this->clients->detach($conn);
    }

    /**
     * If there is an error with one of the sockets, or somewhere in the application where an Exception is thrown,
     * the Exception is sent back down the stack, handled by the Server and bubbled back up the application through this method
     * @param  ConnectionInterface $conn
     * @param  \Exception $e
     * @throws \Exception
     */
    function onError(ConnectionInterface $conn, \Exception $e)
    {
        $conn->close();
    }

    /**
     * Triggered when a client sends data through the socket
     * @param  \Ratchet\ConnectionInterface $from The socket/connection that sent the message to your application
     * @param  string $msg The message received
     * @throws \Exception
     */
    abstract function onMessage(ConnectionInterface $from, $msg);
}
```

**Comment by cboden** [comment]
I don't see any issue with the code you posted, as long as everything is in order with `$routes` and `$container` but you'd probably see other errors if that was the problem.

The `_httpServer` error is perplexing because it's [set on the constructor](https://github.com/ratchetphp/Ratchet/blob/master/src/Ratchet/Http/HttpServer.php#L25) and never changed throughout Ratchet's lifecycle. 

The second error could come of for a few reasons. `$controller` is placed on a Connection by the [`Router`](https://github.com/ratchetphp/Ratchet/blob/master/src/Ratchet/Http/Router.php#L63). Off the top of my head here are a couple ideas that may cause that:
* `$controller` is explicitly `unset` on a `$conn` somewhere
* `HttpServer` is calling `onMessage` of the `Router` before the server has responded

I'll investigate the second one if you can check out the first point in your code base.

**Comment by ikovalyov** [comment]
yes, while investigating code I was thinking about 'HttpServer is calling onMessage of the Router before the server has responded' issue. I can recheck but when I was investigating this issue I didn't find any place with this unset logic.

**Comment by ikovalyov** [comment]
Also, if this can help, these 2 messages are connected - notice first and error after it.

**Comment by cboden** [comment]
Doing a bit of theorizing I believe this _might_ be possible if the client sent an HTTP header request then began streaming a body _before_ the server responded. In addition the Router determined the route was invalid and sent to close the connection. I/O is where order of execution becomes unpredictable so the close header was put in the write buffer but the body onMessage came first. The `HttpServer` called onMessage before the socket finished closing. 
* Client sends HTTP request
* Server deems the route invalid, begins closing
* Client immediately begins streaming body
* Server unexpectedly receives part of body

**Comment by cboden** [comment]
I plan on updating Router and setting a NOOP controller on each connection during `onOpen`. If the above race condition happens there should be no error and the connection will properly close.

**Comment by ikovalyov** [comment]
👍

**Comment by cboden** [comment]
@ikovalyov Can you test your server with the new `router-race-fix` branch and see if you can't get the error to come up?

**Comment by ikovalyov** [comment]
@cboden acrually I don't think it should help. Just checked our logs. Last time this error occured 3 weeks ago.

**Comment by cboden** [comment]
Has any source code changed on your end that may have stopped the error from occurring?

**Comment by ikovalyov** [comment]
no, logic remains the same. but we have many 3rd-party systems connecting

**Comment by cboden** [comment]
(hopefully) fixed by #565  😀
